cdap-app-fabric/src/main/java/co/cask/cdap/metadata/FieldLineageAdmin.java

```diff
@@ -0,0 +1,308 @@
+   * direction specified. Operation details in the "incoming" direction consists of consists of the datasets
+   * and their fields ({@link DatasetField}) that this field originates from, as well as the programs and
+   * operations that generated this field from those origins. In outgoing direction, it consists of the datasets
+   * and their fields ({@link DatasetField}) that were computed from this field, along with the programs and
+   * operations that performed the computation. When direction is specified as 'both', incoming as well
+   * as outgoing operations are returned.
+   *
+   * @param direction the direction can one of the "incoming", "outgoing", or "both"
+   * @param endPointField the EndPointField for which operations to be returned
+   * @param start start time (inclusive) in milliseconds
+   * @param end end time (exclusive) in milliseconds
+   * @return the FieldLineageDetails instance
+   */
+  FieldLineageDetails getOperationDetails(String direction, EndPointField endPointField, long start, long end) {
+    List<ProgramFieldOperationInfo> incoming = null;
+    List<ProgramFieldOperationInfo> outgoing = null;
+    if (direction.equals(Constants.FieldLineage.Direction.INCOMING)
+            || direction.equals(Constants.FieldLineage.Direction.BOTH)) {
+      Set<ProgramRunOperations> incomingOperations = fieldLineageReader.getIncomingOperations(endPointField, start,
+                                                                                              end);
+      incoming = processOperations(incomingOperations);
+    }
+    if (direction.equals(Constants.FieldLineage.Direction.OUTGOING)
+            || direction.equals(Constants.FieldLineage.Direction.BOTH)) {
+      Set<ProgramRunOperations> outgoingOperations = fieldLineageReader.getOutgoingOperations(endPointField, start,
+                                                                                              end);
+      outgoing = processOperations(outgoingOperations);
+    }
+    return new FieldLineageDetails(incoming, outgoing);
+  }
+
+  private List<ProgramFieldOperationInfo> processOperations(Set<ProgramRunOperations> programRunOperations) {
+    List<ProgramFieldOperationInfo> result = new ArrayList<>();
+    for (ProgramRunOperations entry : programRunOperations) {
+      List<ProgramInfo> programInfo = computeProgramInfo(entry.getProgramRunIds());
+      List<FieldOperationInfo> fieldOperationInfo = computeFieldOperationInfo(entry.getOperations());
+      result.add(new ProgramFieldOperationInfo(programInfo, fieldOperationInfo));
+    }
+    return result;
+  }
+
+  /**
+   * Computes the list of {@link ProgramInfo} from given set of ProgramRunIds.
+   * For each program, there is only one entry in the returned list and it is sorted
+   * by the last executed time in descending order.
+   * @param programRunIds set of program run ids from which program info to be computed
```

Comment:
you should always leave a blank line between the description and the @param's

Reply:
fixed.